Second order derivaties for Pooling operations

AnishShah opened the pull request:
Fixes #6143
Fixes #6294 

This is still not a complete. I wanted to know whether this is the correct way? 
I ran the following code and it worked without any error.
If it is correct, then I will add the same for `avg_pool` too.
```
In [1]: import tensorflow as tf
I tensorflow/stream_executor/dso_loader.cc:128] successfully opened CUDA library libcublas.so.7.5 locally
I tensorflow/stream_executor/dso_loader.cc:128] successfully opened CUDA library libcudnn.so.4 locally
I tensorflow/stream_executor/dso_loader.cc:128] successfully opened CUDA library libcufft.so.7.5 locally
I tensorflow/stream_executor/dso_loader.cc:128] successfully opened CUDA library libcuda.so.1 locally
I tensorflow/stream_executor/dso_loader.cc:128] successfully opened CUDA library libcurand.so.7.5 locally

In [2]: x = tf.placeholder(tf.float32, shape=[1, 4, 4, 1])

In [3]: y = tf.nn.max_pool(x, ksize=[1, 2, 2, 1], strides=[1, 2, 2, 1], padding='VALID')

In [4]: dy = tf.gradients(y, x)

In [5]: ddy = tf.gradients(dy, x)
```

AnishShah committed:
[issue #6143] Second order derivaties for Pooling operations

tensorflow-jenkins commented:
Can one of the admins verify this patch?

martinwicke commented:
Jenkins, test this please!

vrv commented on tensorflow/python/ops/nn_grad.py:
```diff
@@ -421,6 +421,13 @@ def _MaxPoolGrad(op, grad):
                                    data_format=op.get_attr("data_format"))
 
 
+@ops.RegisterGradient("MaxPoolGrad")
```
This needs a test, perhaps using tf.test.compute_gradient_error, to prove that this is right :)

AnishShah committed:
add tests and AvgPoolGradGrad

AnishShah commented on tensorflow/python/ops/nn_grad.py:
```diff
@@ -395,6 +395,13 @@ def _AvgPoolGrad(op, grad):
 
+@ops.RegisterGradient("AvgPoolGrad")
+def _AvgPoolGradGrad(op, grad):
+  print(op)
```
@vrv For gradient of gradient of avg_pool, the inputs are input: `gradients/AvgPool_grad/Shape` and `gradients/Fill` which looks wrong. There's no `x` in the input. Is this wrong?

```
In [2]: x = tf.constant(np.random.RandomState(1).rand(1, 4, 4, 1).astype(np.float32), name="x")

In [3]: y = tf.nn.avg_pool(x, ksize=[1, 2, 2, 1], strides=[1, 2, 2, 1], padding='VALID')

In [4]: dy = tf.gradients(y, x)

In [5]: ddy = tf.gradients(dy, x)
name: "gradients/AvgPool_grad/AvgPoolGrad"
op: "AvgPoolGrad"
input: "gradients/AvgPool_grad/Shape"
input: "gradients/Fill"
attr {
  key: "T"
  value {
    type: DT_FLOAT
  }
}
attr {
  key: "data_format"
  value {
    s: "NHWC"
  }
}
attr {
  key: "ksize"
  value {
    list {
      i: 1
      i: 2
      i: 2
      i: 1
    }
  }
}
attr {
  key: "padding"
  value {
    s: "VALID"
  }
}
attr {
  key: "strides"
  value {
    list {
      i: 1
      i: 2
      i: 2
      i: 1
    }
  }
}
```

AnishShah commented:
I have added tests and AvgPoolGrad op. MaxPoolGrad tests are passing but AvgPoolGrad tests are failing. I have added a comment above which might explain why it is failing. @vrv what do you think?

aam-at commented:
The operation is not implemented correctly. If tests are passing, it is only because linear pooling operations are involved which results in all zeros and ones. We need to write a custom gpu kernel for max pool grad grad. Similarly, custom cpu c++ function needs to be implemented. See theano gpu implementation for reference - https://github.com/Theano/Theano/pull/5159. For average pool grad grad, no custom kernel is needed and we can just reuse average pooling.

As I am familiar with theano, I can pick up this pull request fairly easily and add support for max pool grad grad to tensorflow if @AnishShah doesn't mind.

AnishShah commented:
@aam-at Okay. Sure. Please tag me in the PR too if it's okay. I would like to learn how you did it.
Thanks!

AnishShah commented:
@aam-at Also, I would like to know what is wrong?
For the max pool operation, the gradient is passed to the one who has the maximum value inside the kernel window. So, the second order derivative should be zero. This is **kind of** like ReLUs, isn't it? Can you please explain? It would help me understand better. :)

tensorflow-jenkins commented:
Can one of the admins verify this patch?

aam-at commented:
Gradient of max pooling does not always output ones. It will pass top gradients `gz` which correspond to the location of maximum values in the input `x`. If you have only linear operations, then `gz` is ones and `gx` is also ones then. But this is not true in general.

Gradient of the gradient needs to pass gradient of the top input gradient `ggx` which correspond to the locations of maximum values in the input `x`.

For diagram of gradient of max pooling operation, you can look up Zeiler's paper on deconvolutional networks (they have diagram for unpooling which is just basically backward pass of max pooling). Gradient of the gradient is a bit more confusing to show using diagram.

AnishShah commented:
@aam-at Thanks for the explanation. Looking forward to your PR. Please mark me too. :)
